File: light-control/algorithms/two_color_wave/wave.py
import logging
import math
import time
from .config import config
from .._meta import BaseAlgorithm
from .._meta.pixel import Pixel

class Algorithm(BaseAlgorithm):
    def __init__(self, num_pixels, settings) -> None:
        super().__init__(num_pixels, config, settings)
        self.reversed = 1 if settings['reverse'] else -1
        self.speed = settings['speed']
        self.square = 'square' in settings and settings['square']
        
        multiplier = math.tau / settings['wavelength']
        split = 'split' in settings and settings['split']
        if not split:
            self.locs = [n * multiplier for n in range(len(self.pixels))]
        else:
            center = (len(self.pixels) / 2) - 0.5
            self.locs = [math.fabs((n - center) * multiplier) for n in range(len(self.pixels))]

        colors = settings['colors']
        self.base_color = Pixel(colors[0][0], colors[0][1], colors[0][2])
        self.diff = Pixel(colors[1][0], colors[1][1], colors[1][2]).diff(self.base_color)


        self.start_time = 0
        self.run_cycle(0, 0)
        self.start_time = 0
    
    def run_cycle(self, elapsed_millis, elapsed_seconds):
        cur_time = time.time()
        if self.start_time == 0:
            self.start_time = cur_time

        phase = (((cur_time - self.start_time) % self.speed) / self.speed) * self.reversed * math.tau

        for n in range(len(self.pixels)):
            loc = self.locs[n]
            pixel = self.pixels[n]

            pos = (math.cos(loc + phase) + 1) / 2
            if self.square:
                if pos > 0.5: pos = 1
                else: pos = 0
            

            hue = self.diff.hue * pos
            sat = self.diff.sat * pos
            val = self.diff.val * pos

            pixel.hue = self.base_color.hue + hue
            pixel.sat = self.base_color.sat + sat
            pixel.val = self.base_color.val + val

        return super().run_cycle(elapsed_millis, elapsed_seconds)

File: light-control-ts/src/algorithms/twoColorWave/wave.ts
import { BaseAlgorithm, Pixel } from '../_meta/index.js'
import { config } from './config.js'

export class Algorithm extends BaseAlgorithm {
  private readonly reversed: number
  private readonly speed: number
  private readonly square: boolean
  private readonly locs: number[]
  private readonly baseColor: Pixel
  private readonly diff: Pixel
  private startTime = 0

  constructor(numPixels: number, settings: Record<string, unknown>) {
    super(numPixels, config, settings)
    this.reversed = settings['reverse'] ? 1 : -1
    this.speed = settings['speed'] as number
    this.square = !!(settings['square'])

    const multiplier = (2 * Math.PI) / (settings['wavelength'] as number)
    const split = !!(settings['split'])

    if (!split) {
      this.locs = Array.from({ length: this.pixels.length }, (_, n) => n * multiplier)
    } else {
      const center = this.pixels.length / 2 - 0.5
      this.locs = Array.from({ length: this.pixels.length }, (_, n) =>
        Math.abs((n - center) * multiplier),
      )
    }

    const colors = settings['colors'] as number[][]
    this.baseColor = new Pixel(colors[0][0], colors[0][1], colors[0][2])
    this.diff = new Pixel(colors[1][0], colors[1][1], colors[1][2]).diff(this.baseColor)

    this.runCycle(0, 0)
    this.startTime = 0
  }

  runCycle(elapsedMillis: number, elapsedSeconds: number): boolean {
    const curTime = Date.now() / 1000
    if (this.startTime === 0) this.startTime = curTime

    const phase =
      (((curTime - this.startTime) % this.speed) / this.speed) * this.reversed * 2 * Math.PI

    for (let n = 0; n < this.pixels.length; n++) {
      const loc = this.locs[n]
      const pixel = this.pixels[n]

      let pos = (Math.cos(loc + phase) + 1) / 2
      if (this.square) pos = pos > 0.5 ? 1 : 0

      pixel.hue = this.baseColor.hue + this.diff.hue * pos
      pixel.sat = this.baseColor.sat + this.diff.sat * pos
      pixel.val = this.baseColor.val + this.diff.val * pos
    }

    return super.runCycle(elapsedMillis, elapsedSeconds)
  }
}
